File: binary_search/0069_sqrt_x/0069_sqrt_x.py
class Solution(object):
    def mySqrt(self, x):
        if x < 2:
            return x
        left, right = 1, x//2
        while left <= right:
            mid = left +(right - left) // 2
            if mid > x / mid:
                right = mid - 1
            else:
                left = mid + 1

        return left - 1


x = 7
res = Solution().mySqrt(x)
print(res)

File: binary_search/0069_sqrt_x/0069_sqrt_x.go
package main 

import (
	"fmt"
)

func mySqrt(x int) int {
	if x == 0 {
		return 0
	}

	left, right, res := 1, x, 0
	for left <= right {
		mid := left + ((right-left) >> 1)
		if mid < x / mid {
			left = mid + 1
			res = mid
		} else if mid == x / mid {
			return mid
		}else {
			right = mid - 1
		}
	}

	return res
}

func main() {
	x := 4
	res := mySqrt(x)
	fmt.Println(res)
}